src/com/google/enterprise/adaptor/sharepoint/SharePointAdaptor.java

```diff
@@ -3411,6 +3379,62 @@ private HtmlResponseWriter createHtmlResponseWriter(
     public SiteDataClient getSiteDataClient() {
       return siteDataClient;
     }
+      
+    /** 
+     * MemberIdMapping will have information about users with explicit
+     * permissions on SharePoint or users which are direct members of
+     * SharePoint groups. MemberIdMapping might not have information
+     * about all valid SharePoint Users. To get all valid SharePoint users
+     * under SiteCollection, use SiteUserMapping. 
+     */
+    private class IdMappings {
+      private MemberIdMapping memberIdMapping;
+      private MemberIdMapping siteUserMapping;
+      private boolean memberIdMappingRefreshed = false;
+      private boolean siteUserMappingRefreshed = false;
+
+      private IdMappings() throws IOException {
+        this.memberIdMapping = getMemberIdMapping();
+      }
+
+      private Principal resolvePrincipal(Integer id) throws IOException {
+        Principal principal = memberIdMapping.getPrincipal(id);
+        if (principal == null) {
+          log.log(Level.FINE, "Member id {0} is not available in memberid"
+              + " mapping for Web [{1}] under Site Collection [{2}].",
+              new Object[] {id, webUrl, siteUrl});
+          if (siteUserMapping == null) {
+            try {
+              siteUserMapping = getSiteUserMapping();
+              principal = siteUserMapping.getPrincipal(id);
+            } catch (IOException ex) {
+              log.log(Level.FINE, "Could not resolve site user mapping for "
+                  + id, ex);
+            }
+          } else {
+            principal = siteUserMapping.getPrincipal(id);
+          }
+        }
+        if (principal == null && !memberIdMappingRefreshed) {
+          // Try to refresh member id mapping and check again.
+          memberIdMapping = refreshMemberIdMapping(memberIdMapping);
+          memberIdMappingRefreshed = true;
+          principal = memberIdMapping.getPrincipal(id);
+        }
+        if (principal == null && !siteUserMappingRefreshed) {
+          // Try to refresh site user mapping and check again.
+          try {
+            siteUserMapping = refreshSiteUserMapping(siteUserMapping);
```

Comment:
If this throws an exception, we won't consider the siteUserMapping to be refreshed. If it's a persistent error like the bug report, we may retrieve the mapping (since siteUserMapping will remain null) and then refresh it (since siteUserMappingRefreshed will remain false) for up to every entry in the ACL. Is that OK?

My take is that it is not. Inside one call to generateAcl we're in a pretty tight loop, and shouldn't be retrying failing RPC calls on every loop iteration. If the server is down, then something else will fail, or worst case, we miss some for this ACL. If the bug report problem is occurring, we're just thrashing pointlessly.